pkgs/leak_tracker_flutter_testing/lib/src/matchers.dart

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+/// Checks if the object dispatches events to `MemoryAllocations.instance`.
+///
+/// [createAndDispose] should create and dispose the object.
+Matcher dispatchesMemoryEvents(Function() createAndDispose) =>
```

Comment:
I think I like the idea of Matchers instead of helpers even though it might seem odd at first. Helpers obscure important aspects of the test that provide context as someone is reading through a test.
On the other hand matchers provide verbosity and readability, while reading a test, even though some of the work is hidden behind it. 

In this specific case it does feel like [checking if this matcher throws an error](https://github.com/dart-lang/leak_tracker/pull/128/files#diff-def4ff981195fc4483ed7d9a3cc2f6acadd9ac7346fb5a2b5be35a9612f70fd1R37) is a bit of a misuse of the matcher pattern. I wonder if creating a `failsToDispatchMemoryEvents` matcher would be a better way to make that behaviour accessible.

I think precident for using matchers like this has been set already, for example the matcher packages makes behaviour similar to these memory matchers with:
- [emits](https://pub.dev/documentation/matcher/latest/expect/emits.html)
- [emitsError](https://pub.dev/documentation/matcher/latest/expect/emitsError.html)

These emit matchers can be seen in the [stream matchers section](https://pub.dev/packages/matcher#stream-matchers)

Reply:
I'm not saying we shouldn't have a `Matcher` for this, but `Matcher`s should be used to verify results, not actually perform the operations that create the results. In this particular case, the `actual` parameter of `expect(...)` is not the result of an operation but a hard-coded type. The matcher only serves to verify that `actual is Type == true` but then proceeds to do nothing with it before actually running code that should be used to generate `actual` in the first place.

This block of code shouldn't be inside a matcher since it's generating a result rather than verifying one:

```dart
MemoryAllocations.instance.addListener(listener);
createAndDispose();
MemoryAllocations.instance.removeListener(listener);
```

Wrapping that code in a helper that generates a result that can be verified against a `Matcher` isn't any less readable than hiding it in the matcher in the first place.

> In this specific case it does feel like checking if this matcher throws an error is a bit of a misuse of the matcher pattern. I wonder if creating a failsToDispatchMemoryEvents matcher would be a better way to make that behaviour accessible

This can be achieved by wrapping a matcher with `isNot` (e.g., `expect('String', isNot(equal('Foo')))`).

Reply:
hmm now that i'm reading a bit more into this I wonder if I've missunderstood who would be using this. I'm going to loop back on this in an hour after lunch

Reply:
Hmm what if the matcher was built like the [`throwA`](https://pub.dev/packages/matcher#exception-matchers) Matcher. The function could be on the left side of the expect, then we could still do all of the necessary setup and teardown to do the test.

Using the matcher would look a lot more like you expect it:
```dart
expect(
  () => _TrackedClass().dispose(),
  dispatchesMemoryEventsFor(_TrackedClass),
)
expect(
  () => {},
  isNot(dispatchesMemoryEventsFor(_TrackedClass)),
)
```

The Matcher would look something like this instead:
```dart
class _DispatchesMemoryEvents extends Matcher {
  const _DispatchesMemoryEvents(this.type);

  final Type type;
  static const _key = 'description';

  @override
  bool matches(Function block, Map matchState) {
    final events = <ObjectEvent>[];

    void listener(ObjectEvent event) {
      if (event.object.runtimeType == type) {
        events.add(event);
      }
    }

    MemoryAllocations.instance.addListener(listener);
    block();
    MemoryAllocations.instance.removeListener(listener);

    if (events.length == 2 &&
        events.first is ObjectCreated &&
        events.last is ObjectDisposed) {
      return true;
    }

    matchState[_key] =
        'createAndDispose is expected to dispatch two events to $MemoryAllocations.instance,'
        ' for the type $type,'
        ' first $ObjectCreated and then $ObjectDisposed.\n'
        'Actually it dispatched ${events.length} events:\n$events';

    return false;
  }

...
}
```